test/controllers/projects_controller_test.rb

```diff
@@ -55,4 +56,334 @@ class ProjectsControllerTest < ActionDispatch::IntegrationTest
+        ]
+      }
+    }, as: :json
+
+    assert_response :success
+
+    assert_equal 2, @response.parsed_body['issues']['errors'].length
+    # Error should throw if row is missing sample name.
+    assert_match 'Row 0 is missing sample_name.', @response.parsed_body['issues']['errors'][0]
+    # Error should throw if row sample name doesn't exist in this project.
+    assert_match 'foobar (row 1) does not match any sample names in this project.', @response.parsed_body['issues']['errors'][1]
+
+    assert_equal 0, @response.parsed_body['issues']['warnings'].length
+  end
+
+  test 'metadata validate host genome names valid if new samples' do
+    post user_session_path, params: @user_params
+
+    post validate_metadata_csv_project_url(@metadata_validation_project), params: {
+      metadata: {
+        headers: ['host_genome_name', 'sample_type'],
+        rows: [
+          ['', 'Whole Blood'],
+          ['foobar', 'Whole Blood'],
+          ['Human', 'Whole Blood']
+        ]
+      },
+      new_samples: '1'
+    }, as: :json
+
+    assert_response :success
+
+    assert_equal 2, @response.parsed_body['issues']['errors'].length
+    # Error should throw if row is missing host genome name.
+    assert_match 'Row 0 is missing host_genome_name.', @response.parsed_body['issues']['errors'][0]
+    # Error should throw if host genome name is invalid.
+    assert_match 'foobar (row 1) is an invalid host_genome_name.', @response.parsed_body['issues']['errors'][1]
+
+    assert_equal 0, @response.parsed_body['issues']['warnings'].length
+  end
+
+  test 'metadata validate values valid' do
+    post user_session_path, params: @user_params
+
+    post validate_metadata_csv_project_url(@metadata_validation_project), params: {
+      metadata: {
+        headers: ['sample_name', 'sample_type', 'age', 'admission_date', 'blood_fed', 'reported_sex'],
+        rows: [
+          ['metadata_validation_sample_human', 'Whole Blood', 100, '2018-01-01', '', ''],
+          ['metadata_validation_sample_human', 'Whole Blood', 'foobar', 'foobar', 'foobar', 'foobar'],
+          ['metadata_validation_sample_mosquito', 'Whole Blood', '', '', 'Yes', 'Male'],
+          ['metadata_validation_sample_mosquito', 'Whole Blood', 'foobar', 'foobar', 'foobar', 'foobar']
+        ]
+      }
+    }, as: :json
+
+    assert_response :success
+
+    assert_equal 7, @response.parsed_body['issues']['errors'].length
+    # Error should throw if invalid float is passed for float data type.
+    assert_match 'foobar is not a valid number (row 1)', @response.parsed_body['issues']['errors'][0]
+    # Error should throw if invalid date is passed for date data type.
+    assert_match 'foobar is not a valid date (row 1)', @response.parsed_body['issues']['errors'][1]
+    # Error should throw if metadata type is not supported for the sample's host genome.
+    assert_match 'blood_fed is not a supported metadatum for host genome Human (row 1)', @response.parsed_body['issues']['errors'][2]
```

Comment:
Slightly worried that it will be a pain to fix all the tests if we want to change the wording of any of these error messages... Maybe extract the boilerplate text out into functions?

Reply:
Done.